### PR TITLE
Dlib 19.0

### PR DESCRIPTION
**Pull request by patricksnape**
Currently struggling with ~~BLAS on Unix and~~ JPEG on Windows. Let's see what happens on the CI.

**Patches explanation:**

  - Davis forces building JPEG on OSX - this is removed. Will address upstream when confirmed working.
  - Davis forces static linking of boost on Windows *and* always looks for Python 3.4. Fixed by removing checks and using the `%PY_VER%` environment variable. Will address upstream when confirmed working.
  - The function check fails on Windows for `libpng`, so I remove it to ensure that we link against `libpng`
  - The function checks fails for `openblas` on Unix, so I remove it to ensure that we link against `openblas`

**Issues seen locally:**

  - Runtime error on Windows when loading JPEG images
  - ~~OSX linking error against new openblas~~ Not issue on CI
  - OSX still linking to system (Accelerate) LAPACK
    - How can I test this on CI? Run `otool -L`?
  - ~~Linux still linking to `mkl` despite it not seeming to be mentioned anywhere - though this may be a quirk of my box.~~ Not issue on CI

### CODE REVIEW

**Comment by patricksnape** [comment]
@jakirkham There is some weird bug whereby the build environment isn't getting cleaned up properly. Not sure if this is `conda-build-all` or `conda-build`. See [here](https://travis-ci.org/conda-forge/staged-recipes/jobs/149155097) and [here](https://circleci.com/gh/conda-forge/staged-recipes/6811) where it complains about the `build` directory already existing when starting the *second* build for Python 3.4.

I have worked around this by adding an `rm -fr` for now, but this seems like a bug.

**Comment by patricksnape** [comment]
CI is lying due to that last skip commit - failing completely on Windows due to jpeg. @davisking just wondering if you had any advise on how I might debug this - have you ever actually seen jpeg loading work on Windows for dlib 19.0?

This is my first time building dlib 19.0 so I'm a bit frustrated my old recipe didn't "just work".

**Comment by patricksnape** [comment]
@183amir Dlib recipe is here but I'm struggling with Windows builds at the moment.

**Comment by 183amir** [comment]
Is Dlib a requirement of `cyvlfeat`? I don't remember needing Dlib.

**Comment by davisking** [comment]
I haven't tried to load any jpeg files on windows using dlib since the last release.  I would be surprised if it suddenly wasn't working though.  Are you building using the copy of libjpeg in dlib/external or trying to link against some other libjpeg?  What error do you get?

**Comment by patricksnape** [comment]
@183amir sorry for the noise - I got confused and thought you had requested dlib previously.

@davisking I've tried both. If I try linking to the external jpeg then I get a linking error (not to mention I have to do some horrible hacking around defines inside `windows.h`. This recipe at the moment is just trying to use the jpeg (unaltered) from inside external. The error is as follows:

```
======================================================================
ERROR: test_train_xml_detector (__main__.TestDlib)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Miniconda\conda-bld\test-tmp_dir\run_test.py", line 116, in test_train_xml_detector
    dlib.train_simple_object_detector('images.xml', "test.svm", options)
RuntimeError: Unable to load image in file face.jpg.
You must #define DLIB_JPEG_SUPPORT and link to libjpeg to read JPEG files.
Do this by following the instructions at http://dlib.net/compile.html.
Note that you must cause DLIB_JPEG_SUPPORT to be defined for your entire project.
So don't #define it in one file. Instead, add it to the C/C++->Preprocessor->Preprocessor Definitions field in Visual Studio's Property Pages window so it takes effect for your entire application.
```
and the test code is
```
def test_train_xml_detector(self):
    # This effectively tests that we can successfully load images
    options = dlib.simple_object_detector_training_options()
    options.add_left_right_image_flips = True
    options.C = 1
    options.num_threads = 1

    dlib.train_simple_object_detector('images.xml', "test.svm", options)
    self.assertTrue(os.path.exists('./test.svm'))
```

**Comment by davisking** [comment]
Apparently DLIB_JPEG_SUPPORT isn't defined :)

Are you using dlib's cmake scripts?  They should do that automatically.

File: recipes/dlib/bld.bat [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+-DPYTHON_LIBRARY="%PREFIX%\libs\python%PY_VER_NO_DOT%.lib"  ^
+-DPYTHON_INCLUDE_DIR="%PREFIX%\include"                     ^
+-DDLIB_LINK_WITH_SQLITE3=0                                  ^
+-DDLIB_PNG_SUPPORT=1                                        ^
+-DPNG_INCLUDE_DIR="%LIBRARY_INC%"                           ^
+-DPNG_PNG_INCLUDE_DIR="%LIBRARY_INC%"                       ^
+-DPNG_LIBRARY=%PNG_LIBRARY%                                 ^
+-DZLIB_INCLUDE_DIRS="%LIBRARY_INC%"                         ^
+-DZLIB_LIBRARIES="%LIBRARY_BIN%\zlib.dll"                   ^
+-DDLIB_JPEG_SUPPORT=1                                       ^
```
Comment (patricksnape) [inline recipes/dlib/bld.bat]: @davisking But it is defined!

Comment (davisking) [inline recipes/dlib/bld.bat]: Beats me :)   It's definitely not defined though.  There must be some
problem somewhere in the build system.  Or maybe that command is too long
for windows and it truncates it.

Comment (patricksnape) [inline recipes/dlib/bld.bat]: [L503](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.5184/job/iiixxpbb1ta6fe7m#L503) suggests it is indeed set (plus it actually [builds jpeg](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.5184/job/iiixxpbb1ta6fe7m#L803). Were you always setting `CXX_FLAGS` on Windows to add the defines (rather than `add_definitions`)?

Comment (davisking) [inline recipes/dlib/bld.bat]: That looks right.  Yeah, I always use that add_global_define cmake macro
because otherwise parent cmake projects don't see the #define.

**Comment by davisking** [comment]
I just ran into this when testing on windows.  I'm using cmake 3.6.0-rc3 and when I generate a project it doesn't set DLIB_JPEG_SUPPORT (or any other defines for that matter).  But if I just run cmake again then it sets them.  I haven't seen this behavior with older versions of cmake.  It seems like this is going to be a problem though.  I'll see what can be done.  But let me know if you find a solution first :)

**Comment by davisking** [comment]
I tried a bunch of things last night.  It turns out that if you put the
flags into CMAKE_CXX_FLAGS, which is what the dlib scripts do, you get this
problem.  But if the flags are added instead to CMAKE_CXX_FLAGS_RELEASE or
any of the other modal versions of this variable it works the first time.
But then when you try to toggle options in cmake-gui that would effect
these flags they aren't updated correctly.  It's almost like the reverse
problem.

I haven't figured out a work around yet :/

**Comment by patricksnape** [comment]
@davisking OK well that at least allows me to patch here - now I've got some annoying build bug on Python 3.4 - but that is a new problem!

**Comment by patricksnape** [comment]
I don't see the failures locally on Windows so open and close to try rebuilding?

**Comment by patricksnape** [comment]
I don't see this linking error locally - though I have VS 2010 Ultimate rather than Express. Next step is to turn verbose building on to see if we can get more information about the linking error - since NMake eats the error.

**Comment by patricksnape** [comment]
Uhh... apparently this just passed... @jakirkham a review please?

**Comment by jakirkham** [comment]
Nice work!

Alright I'll take a look. Might take me a day or two to get to though. Hope that is ok.

File: recipes/dlib/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  requires:
+    - tqdm
+    - pillow
+
+  imports:
+    - dlib
+
+about:
+  home: http://dlib.net/
+  license: Boost Software License
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Please add `license_file: dlib/LICENSE.txt` to include the license file.

File: recipes/dlib/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  features:  # [not win]
+    - blas_{{ blas_variant }}  # [not win]
+
+requirements:
+  build:
+   - python
+   - cmake
+   - toolchain
+   - openblas              # [not win]
+   - numpy >=1.10
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: It appears we are linking against `numpy`. Please also add `numpy x.x`.

File: recipes/dlib/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+   - boost 1.61.*
+   - jpeg 9*               # [not win]
+   - libpng >=1.6.21,<1.7
+   - sqlite 3.13.*         # [not win]
+
+  run:
+   - python
+   - openblas              # [not win]
+   - numpy >=1.10
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Please also add `numpy x.x`.

File: recipes/dlib/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+source:
+  url: https://github.com/davisking/{{ name }}/archive/v{{ version }}.tar.gz
+  sha256: a6463524cdb1d275ab408f182b0d47af143179fde7f1dd97074dc25f2f9a2676
+
+  patches:
+    - osx_jpeg.patch                # [osx]
+    - lapack_sgetrf.patch           # [unix]
+    - win_boost.patch               # [win]
+    - win_png.patch                 # [win]
+    - add_global_cxx_release.patch
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Have these patches been sent upstream? If so, it would be good to cite the PR/issue link a comment in the recipe. Possibly with a small explanation about each one.

File: recipes/dlib/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  number: 0
+  features:  # [not win]
+    - blas_{{ blas_variant }}  # [not win]
+
+requirements:
+  build:
+   - python
+   - cmake
+   - toolchain
+   - openblas              # [not win]
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Needs to be pinned to `0.2.18|0.2.18.*`.

File: recipes/dlib/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+   - numpy >=1.10
+   - boost 1.61.*
+   - jpeg 9*               # [not win]
+   - libpng >=1.6.21,<1.7
+   - sqlite 3.13.*         # [not win]
+
+  run:
+   - python
+   - openblas              # [not win]
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Needs to be pinned to `0.2.18|0.2.18.*`.

**Comment by jakirkham** [comment]
Should note pinnings live [here]( https://github.com/conda-forge/conda-forge.github.io/blob/6b51a708414e7db952b9b821a3601e675c814495/scripts/pin_the_slow_way.py#L29-L53 ) now.

File: recipes/dlib/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Also a license version would be good. We can probably shorten this to `Boost`. So it would be something like `Boost 1.0` should be fine.

File: recipes/dlib/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    - tqdm
+    - pillow
+
+  imports:
+    - dlib
+
+about:
+  home: http://dlib.net/
+  license: Boost Software License
+  summary: Header only machine learning toolkit
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Seems like it some `cpp` files at least for binding to Python if nothing else. Maybe we can use the repo description line or similar.

File: recipes/dlib/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+build:
+  number: 0
+  features:  # [not win]
+    - blas_{{ blas_variant }}  # [not win]
+
+requirements:
+  build:
+   - python
+   - cmake
+   - toolchain
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Please add `- blas 1.1 {{ variant }}`.

File: recipes/dlib/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+   - openblas              # [not win]
+   - numpy >=1.10
+   - boost 1.61.*
+   - jpeg 9*               # [not win]
+   - libpng >=1.6.21,<1.7
+   - sqlite 3.13.*         # [not win]
+
+  run:
+   - python
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Please add `- blas 1.1 {{ variant }}`.

**Comment by jakirkham** [comment]
So what happens on Windows if you don't have a BLAS?

File: recipes/dlib/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+  patches:
+    - osx_jpeg.patch                # [osx]
+    - lapack_sgetrf.patch           # [unix]
+    - win_boost.patch               # [win]
+    - win_png.patch                 # [win]
+    - add_global_cxx_release.patch
+
+build:
+  number: 0
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Please set the build number to `200`. Though I suppose this is only a requirement where a BLAS is present.

**Comment by jakirkham** [comment]
This has the [info]( https://conda-forge.hackpad.com/BLAS-Numpy-Friends-86De62hoNdT ) about how we setup BLAS to work.

File: recipes/dlib/bld.bat [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+-DDLIB_USE_BLAS=0                                           ^
+-DDLIB_USE_LAPACK=0                                         ^
+-DUSE_SSE2_INSTRUCTIONS=1                                   ^
+-DUSE_SSE4_INSTRUCTIONS=0                                   ^
+-DUSE_AVX_INSTRUCTIONS=0                                    ^
```
Comment (jakirkham) [inline recipes/dlib/bld.bat]: Are these runtime checked?

File: recipes/dlib/build.sh [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  -DZLIB_INCLUDE_DIRS="$PREFIX/include"                 \
+  -DZLIB_LIBRARIES="$PREFIX/lib/libz.${DYNAMIC_EXT}"    \
+  -DDLIB_JPEG_SUPPORT=1                                 \
+  -DJPEG_INCLUDE_DIR="$PREFIX/include"                  \
+  -DJPEG_LIBRARY="$PREFIX/lib/libjpeg.${DYNAMIC_EXT}"   \
+  -DDLIB_LINK_WITH_SQLITE3=1                            \
+  -DDLIB_NO_GUI_SUPPORT=1                               \
+  -DUSE_SSE2_INSTRUCTIONS=1                             \
+  -DUSE_SSE4_INSTRUCTIONS=$USE_SSE4                     \
+  -DUSE_AVX_INSTRUCTIONS=0                              \
```
Comment (jakirkham) [inline recipes/dlib/build.sh]: Same question as above. Are these runtime determined?

File: recipes/dlib/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    - dlib
+
+about:
+  home: http://dlib.net/
+  license: Boost Software License
+  summary: Header only machine learning toolkit
+
+extra:
+  recipe-maintainers:
+    - patricksnape
```
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: Any interest in being added as a maintainer, @davisking? 😉

Comment (davisking) [inline recipes/dlib/meta.yaml]: That's OK.  I have enough open source stuff to maintain already :)

File: recipes/dlib/build.sh (hunk 13 above)
Comment (patricksnape) [inline recipes/dlib/build.sh]: No they are not.

File: recipes/dlib/add_global_cxx_release.patch [hunk 15]
```diff
@@ -0,0 +1,27 @@
+--- dlib/add_global_compiler_switch.cmake
++++ dlib/add_global_compiler_switch.cmake
+@@ -7,18 +7,18 @@
+ macro ( add_global_compiler_switch switch_name )
+    # If removing the switch would change the flags then it's already present
+    # and we don't need to do anything.
+-   string(REPLACE "${switch_name}" "" tempstr "${CMAKE_CXX_FLAGS}")
+-   if ("${CMAKE_CXX_FLAGS}" STREQUAL "${tempstr}" )
+-      set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${switch_name}" 
++   string(REPLACE "${switch_name}" "" tempstr "${CMAKE_CXX_FLAGS_RELEASE}")
```
Comment (patricksnape) [inline recipes/dlib/add_global_cxx_release.patch]: @jakirkham This is currently broken in Dlib master - but it's kind of a nasty hack that only makes it work for RELEASE mode. It definitely needs fixing upstream but I don't have the bandwidth to come up with a 'proper' fix at the moment.

File: recipes/dlib/lapack_sgetrf.patch [hunk 16]
```diff
@@ -0,0 +1,20 @@
+--- dlib/cmake_find_blas.txt
++++ dlib/cmake_find_blas.txt
+@@ -140,15 +140,8 @@
+          set(blas_found 1)
+          message(STATUS "Found OpenBLAS library")
+          set(CMAKE_REQUIRED_LIBRARIES ${blas_libraries})
+-         # If you compiled OpenBLAS with LAPACK in it then it should have the
```
Comment (patricksnape) [inline recipes/dlib/lapack_sgetrf.patch]: This just didn't seem to work? But I'm not sure if it was a quirk of our OpenBLAS build or not. The function `sgetrf_single` is in the binary, but it *is* marked as private - so this fails. I don't think that OpenBLAS is being used a lot with dlib. Again, it would be up to @davisking if he wanted to remove this check in general.

File: recipes/dlib/osx_jpeg.patch [hunk 17]
```diff
@@ -0,0 +1,12 @@
+--- dlib/CMakeLists.txt
++++ dlib/CMakeLists.txt
+@@ -362,7 +362,7 @@
+             set(CMAKE_REQUIRED_LIBRARIES ${JPEG_LIBRARY})
+             CHECK_FUNCTION_EXISTS(jpeg_read_header LIBJPEG_IS_GOOD)
+          endif()
+-         if (JPEG_FOUND AND LIBJPEG_IS_GOOD AND NOT APPLE)
++         if (JPEG_FOUND AND LIBJPEG_IS_GOOD)
```
Comment (patricksnape) [inline recipes/dlib/osx_jpeg.patch]: This can actually probably be put upstream because I don't think this is an issue for modern OSX machines - unless we try and check for the OSX build before disabling it automatically.

File: recipes/dlib/win_boost.patch [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+@@ -34,37 +34,16 @@
+ get_filename_component(PYTHON_PATH ${PYTHON_PATH} PATH)
+ set(CMAKE_PREFIX_PATH ${PYTHON_PATH})
+ 
+-
+-# To avoid dll hell, always link everything statically when compiling in
+-# visual studio.  This way, the resulting library won't depend on a bunch
+-# of other dll files and can be safely copied to someone elese's computer
+-# and expected to run.
+-if (MSVC)
```
Comment (patricksnape) [inline recipes/dlib/win_boost.patch]: This is a choice @davisking kind made that we don't want to replicate. Other option would be to add extra flags to disable/enable this through user choice.

File: recipes/dlib/win_boost.patch [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+-        FIND_PACKAGE(Boost 1.41.0 COMPONENTS python)
+-    endif()
+-    FIND_PACKAGE(PythonLibs 3.4 REQUIRED)
++if (PYTHON3)  
++    FIND_PACKAGE(Boost 1.41.0  COMPONENTS python3 )
++    FIND_PACKAGE(PythonLibs $ENV{PY_VER} REQUIRED)
+ else()
+     FIND_PACKAGE(Boost 1.41.0 COMPONENTS python)
+-    FIND_PACKAGE(PythonLibs 2.6 REQUIRED)
++    FIND_PACKAGE(PythonLibs $ENV{PY_VER} REQUIRED)
```
Comment (patricksnape) [inline recipes/dlib/win_boost.patch]: This is conda specific. Although the more general bug is the explicit seeking of Python 3.4 (which kills 3.5 builds). Should probably be fixed upstream.

File: recipes/dlib/win_png.patch [hunk 20]
```diff
@@ -0,0 +1,51 @@
+--- dlib/CMakeLists.txt
++++ dlib/CMakeLists.txt
+@@ -303,48 +303,9 @@
+          # installed on this system.  
+          if (PNG_FOUND)
+             set(CMAKE_REQUIRED_LIBRARIES ${PNG_LIBRARY})
+-            CHECK_FUNCTION_EXISTS(png_create_read_struct LIBPNG_IS_GOOD)
```
Comment (patricksnape) [inline recipes/dlib/win_png.patch]: This was failing and I have no idea why - so just disabling it made it link properly.

**Comment by patricksnape** [comment]
@jakirkham Thanks for the review - rebased and made the changes. Will have to wait for CI to pass and then we are good to go. Only issue is that some of the patches were to fix discrepancies between how Davis had created the CMake file and how we want to build things in conda-forge. Furthermore, a lot of the build system has changed after 0.19 and so the upstream patches would make no sense! Are we hard on those upstream patches otherwise this may take a very long while to get in as I have no bandwidth to make 'proper' fixes at the moment?

File: recipes/dlib/win_png.patch (hunk 20 above)
Comment (jakirkham) [inline recipes/dlib/win_png.patch]: Knowing `libpng` it is probably some kind of ABI break. 😩

**Comment by jakirkham** [comment]
Yeah, build changes are the most challenging kind of patch to go upstream for sure. My recommendation would simply raise an issue upstream about each of these patches (assuming they haven't been somehow addressed already). In any event, I'm happy to see this merged before that happens. Thanks @patricksnape.

File: recipes/dlib/meta.yaml (hunk 14 above)
Comment (jakirkham) [inline recipes/dlib/meta.yaml]: No worries. Thanks for your help in getting this built. 😄

**Comment by patricksnape** [comment]
Thanks a lot @jakirkham !

File: recipes/dlib/lapack_sgetrf.patch (hunk 16 above)
Comment (davisking) [inline recipes/dlib/lapack_sgetrf.patch]: Does it actually work with that copy of openblas that is missing symbols?
I've added checks like that specifically because some systems include
broken copies of various libraries which eventually cause linker errors due
to missing symbols.  Does that openblas actually have lapack in it?

Comment (jakirkham) [inline recipes/dlib/lapack_sgetrf.patch]: > Does that openblas actually have lapack in it?

Yep, and I use it regularly with NumPy, SciPy and a few other LAPACK dependent libraries.

Comment (jakirkham) [inline recipes/dlib/lapack_sgetrf.patch]: Can't speak to whether dlib works with it or not ATM. Though we should be able to check soon.

File: recipes/dlib/win_boost.patch (hunk 18 above)
Comment (davisking) [inline recipes/dlib/win_boost.patch]: I get unending amounts of email from users who do not, in any sense,
understand the distinction between static and dynamic linking, or even what
linking is at all.  If the default in dlib's files isn't to statically link
I will get even more emails about "dlib runtime errors" that are because
people have dynamically linked to dlls that are not available when they run
python because they have misconfigured their systems, not because of any
dlib related problems.  So I'm not super excited to change this.

Comment (patricksnape) [inline recipes/dlib/win_boost.patch]: Yeah I totally understand! Hence why I just made a patch here that just ripped that out. I'm not sure you should change it to be honest given the complicated build setup required for dlib.

File: recipes/dlib/lapack_sgetrf.patch (hunk 16 above)
Comment (patricksnape) [inline recipes/dlib/lapack_sgetrf.patch]: @davisking We control the OpenBLAS build so if it's broken then we can fix it/that is a bug. Are there any parts of dlib exposed via the Python API that make heavy use of BLAS/LAPACK?

**Comment by davisking** [comment]
Ha, well, it's not that dlib has complicated build requirements.  You can
build it with all this stuff off and it will work fine.  I've just tried to
make the cmake scripts build sensible programs on all platforms for all
people.  For example, for a long time dlib didn't include libpng or say
anything about it.  But I was constantly emailed by people who couldn't
figure out what "link to libpng if you want to use it" means.  So then I
made cmake do that for you.  But then there are people who would email,
"I'm trying to open a png and it doesn't work" because they don't have
libpng installed and don't understand how to install it.  So now dlib
includes a copy of libpng in the zip file and cmake will automatically
compile it if it doesn't find a copy on your system.  Then I found out,
from more emails, that there are lots of people who don't understand any of
this stuff at any level and also have broken copies of libpng installed
that cmake thinks it can use.  So now there are those checks in the cmake
scripts to try and avoid broken libpng libraries.

All of this doesn't have anything to do with dlib :).

I always tell people to use cmake because it's pretty easy and automatic in
most cases.  Moreover, I get regular emails from people who say things like
"I'm using Visual Studio/XCode/SomeRandomIDE and I don't understand what
you mean by 'add the .cpp files to my project'".  So I just tell them to
run the cmake commands.

However, you guys might want to just avoid the dlib cmake scripts
altogether if you are trying to package copies of all these other
libraries. All you need to do is compile dlib/all/source.cpp and the .cpp
files in tools/python/src, then link to any additional stuff you want to
use like libpng.  I don't think the python bindings even make use most of
the optional dependencies like libpng though, with exception of BLAS a
little bit.  Really you are probably fine if you don't tell dlib to link to
anything as far as the python API goes.

**Comment by davisking** [comment]
@patricksnape, If you aren't getting errors I wouldn't worry about it.  The
python API hardly uses BLAS and LAPACK.
